### PR TITLE
feat: add HMR plugin

### PR DESCRIPTION
**Pull request by Suyashtnt**
This adds an HMR plugin, which can be used to automatically remove, add, and update `Piece`s based on file changes 

---

Edit by maintainer:

Closes sapphiredev/framework#353

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -103,4 +103,4 @@
 	},
 	"prettier": "@sapphire/prettier-config",
 	"packageManager": "yarn@3.1.1"
-}
+}
```
Comment (kyranet) [inline package.json:106]: Why was this changed?

File: packages/hmr/src/register.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+		this.logger.info('HMR is enabled!');
+		for (const store of this.stores.values()) {
+			const deleteCb = handlePiecePathDelete.bind(null, store);
+			const updateCb = handlePiecePathUpdate.bind(null, store);
+
+			chokidar
+				.watch([...store.paths])
+				.on('change', updateCb)
+				.on('unlink', deleteCb);
+		}
```
Comment (kyranet) [inline packages/hmr/src/register.ts:24]: Old code, should be moved to `start()`

**Review by kyranet — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (Suyashtnt) [reply package.json:106]: ask jetbrains not me. Ill fix it

**Comment by Suyashtnt** [comment]
example of it working:
https://cdn.discordapp.com/attachments/737142940777971773/928257249808351242/parsecd_2022.01.05_-_14.01.31.35.mp4

File: package.json (hunk 1 above)
Comment (kyranet) [reply package.json:106]: Still not fixed.

File: packages/hmr/README.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+```typescript
+if (process.env.NODE_ENV === 'development') import('@sapphire/plugin-hmr/register');
+```
+
+### TypeScript
+
+in your main or setup file, register the plugin:
+
+```typescript
+import('@sapphire/plugin-hmr/register');
```
Comment (kyranet) [inline packages/hmr/README.md:64]:
```suggestion
import '@sapphire/plugin-hmr/register';
```

File: packages/hmr/README.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+in your main or setup file, register the plugin:
+
+```javascript
+require('@sapphire/plugin-hmr/register');
+```
+
+if you want to make sure its only loaded in development, you can do this:
+
+```typescript
+if (process.env.NODE_ENV === 'development') import('@sapphire/plugin-hmr/register');
```
Comment (kyranet) [inline packages/hmr/README.md:56]: Should include the docs for `start()` as exported from the index.

File: packages/hmr/README.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+in your main or setup file, register the plugin:
+
+```typescript
+import('@sapphire/plugin-hmr/register');
+```
+
+if you want to make sure its only loaded in development, you can do this:
+
+```typescript
+if (process.env.NODE_ENV === 'development') import('@sapphire/plugin-hmr/register');
```
Comment (kyranet) [inline packages/hmr/README.md:70]: Ditto.

**Review by kyranet — CHANGES_REQUESTED** [review]

**Review by kyranet — CHANGES_REQUESTED** [review]
A few

File: packages/hmr/README.md [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+```typescript
+import '@sapphire/plugin-hmr/register';
+```
+
+if you want to make sure its only loaded in development, you can do this:
+
+```typescript
+// at the top of your file
+import { start } from '@sapphire/plugin-hmr';
```
Comment (kyranet) [inline packages/hmr/README.md:73]: Why does this have a comment while the other doesn't? And it's unneeded.

```suggestion
import { start } from '@sapphire/plugin-hmr';
```

File: packages/hmr/README.md [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+```javascript
+require('@sapphire/plugin-hmr/register');
+```
+
+if you want to make sure its only loaded in development, you can do this:
+
+```javascript
+const { start } = require('@sapphire/hmr');
+
+if (process.env.NODE_ENV === 'development') start();
```
Comment (kyranet) [inline packages/hmr/README.md:58]: Clarify with a comment that this needs to be called after `SapphireClient#login`, so it registers the hooks for all the stores and all their paths (user paths are registered in login).

Perhaps something like:

```javascript
// Define client and everything

async function main() {
	await client.login();

	if (process.env.NODE_ENV === 'development') start();
}

void main();
```

**Review by kyranet — CHANGES_REQUESTED** [review]

File: README.md [hunk 8]
```diff
@@ -18,14 +18,17 @@
...
 
 ## Buy us some doughnuts
 
-Sapphire Community is and always will be open source, even if we don't get donations. That being said, we know there are amazing people who may still want to donate just to show their appreciation. Thank you very much in advance!
+Sapphire Community is and always will be open source, even if we don't get donations. That being said, we know there are
+amazing people who may still want to donate just to show their appreciation. Thank you very much in advance!
 
-We accept donations through Open Collective, Ko-fi, Paypal, Patreon and GitHub Sponsorships. You can use the buttons below to donate through your method of choice.
+We accept donations through Open Collective, Ko-fi, Paypal, Patreon and GitHub Sponsorships. You can use the buttons
+below to donate through your method of choice.
```
Comment (favna) [inline README.md:31]: Revert

File: README.md [hunk 9]
```diff
@@ -76,4 +79,5 @@ Thanks goes to these wonderful people ([emoji key](https://allcontributors.org/d
 
 <!-- ALL-CONTRIBUTORS-LIST:END -->
 
-This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification. Contributions of any kind welcome!
+This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification.
+Contributions of any kind welcome!
```
Comment (favna) [inline README.md:83]: Revert

File: packages/hmr/README.md [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+  </tr>
+</table>
+
+<!-- markdownlint-restore -->
+<!-- prettier-ignore-end -->
+
+<!-- ALL-CONTRIBUTORS-LIST:END -->
+
+This project follows the [all-contributors](https://github.com/all-contributors/all-contributors) specification.
+Contributions of any kind welcome!
```
Comment (favna) [inline packages/hmr/README.md:179]: Dump the content from `/README.md` in here after reverting the changes as requested

File: packages/hmr/README.md [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+[![codecov](https://codecov.io/gh/sapphiredev/plugins/branch/main/graph/badge.svg?token=QWL8FB16BR)](https://codecov.io/gh/sapphiredev/plugins)
+[![npm bundle size](https://img.shields.io/bundlephobia/min/@sapphire/plugin-hmr?logo=webpack&style=flat-square)](https://bundlephobia.com/result?p=@sapphire/plugin-hmr)
+[![npm](https://img.shields.io/npm/v/@sapphire/plugin-hmr?color=crimson&logo=npm&style=flat-square)](https://www.npmjs.com/package/@sapphire/plugin-hmr)
+
+</div>
+
+## Description
+
+This plugin provides a HMR (Hot Module Replacement) feature for @sapphire/framework. This allows you to add, delete, and
+update commands, listeners, and more without having to restart your bot. This allows for rapid iteration and prototyping
```
Comment (favna) [inline packages/hmr/README.md:19]:
```suggestion
update commands, listeners, and other pieces without having to restart your bot. This allows for rapid iteration and prototyping
```

File: packages/hmr/README.md [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+without your bot slowing you down.
+
+## Features
+
+-   Fully ready for TypeScript!
+-   Includes ESM ready entrypoint
+-   FAST!
+-   Super easy to use
```
Comment (favna) [inline packages/hmr/README.md:27]:
```suggestion
-   Fully ready for TypeScript!
-   Includes ESM ready entrypoint
-   Easy to use
```

File: packages/hmr/README.md [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+npm install @sapphire/plugin-hmr @sapphire/framework
+```
+
+---
+
+## Usage
+
+### JavaScript
+
+in your main or setup file, register the plugin:
```
Comment (favna) [inline packages/hmr/README.md:47]:
```suggestion
In your main or setup file, register the plugin:
```

File: packages/hmr/README.md [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+```javascript
+require('@sapphire/plugin-hmr/register');
+```
+
+if you want to make sure its only loaded in development, you can do this:
```
Comment (favna) [inline packages/hmr/README.md:53]:
```suggestion
Or if you want to make sure the plugin is only loaded in development, you can register it dynamically like so:
```

File: packages/hmr/README.md [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+
+const client = new SapphireClient({
+	/* your bot options */
+});
+
+async function main() {
+	await client.login();
+
+	// this has to be called after you have logged into your bot
+	if (process.env.NODE_ENV === 'development') start();
```
Comment (favna) [inline packages/hmr/README.md:66]:
```suggestion
	if (process.env.NODE_ENV === 'development') {
		start();
	}
```

File: packages/hmr/README.md [hunk 16]
```diff
@@ -0,0 +1,179 @@
...
+}
+
+main();
+```
+
+### TypeScript
+
+in your main or setup file, register the plugin:
```
Comment (favna) [inline packages/hmr/README.md:74]:
```suggestion
In your main or setup file, register the plugin:
```

File: packages/hmr/README.md [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+
+```typescript
+import '@sapphire/plugin-hmr/register';
+```
+
+if you want to make sure its only loaded in development, you can do this:
```
Comment (favna) [inline packages/hmr/README.md:80]:
```suggestion
Or if you want to make sure the plugin is only loaded in development, you can register it dynamically like so:
```

File: packages/hmr/README.md [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+
+const client = new SapphireClient({
+	/* your bot options */
+});
+
+async function main() {
+	await client.login();
+
+	// this has to be called after you have logged into your bot
+	if (process.env.NODE_ENV === 'development') start();
```
Comment (favna) [inline packages/hmr/README.md:93]:
```suggestion
	if (process.env.NODE_ENV === 'development') {
		start();
	}
```

File: packages/hmr/README.md [hunk 19]
```diff
@@ -0,0 +1,179 @@
...
+void main();
+```
+
+In order for proper HMR, you will need to recompile your code, so you will need to edit your package.json:
+
+- replace the dev and start script 
+```json
+"scripts": {
+	"dev": "run-p watch:build start",
+	"start": "run-s build start:bot"
```
Comment (favna) [inline packages/hmr/README.md:105]:
```suggestion
	"dev": "run-p watch start",
	"build": "tsc",
	"watch": "tsc --watch",
	"start": "node dist/index.js"
```

File: packages/hmr/README.md [hunk 20]
```diff
@@ -0,0 +1,179 @@
...
+"scripts": {
+	"start:bot": "node dist/index.js"
+}
+```
+
+then run your bot using the dev script. You can replace npm with your package manager of choice.
+```sh
+npm run start
+```
+
```
Comment (favna) [inline packages/hmr/README.md:120]:
```suggestion
_Note:_ This uses the `run-p` script which is part of [`npm-run-all`](https://github.com/mysticatea/npm-run-all)
_Note 2:_ Please do note that because the processes are started simultaneously you should run `build` at least once before running `dev`, otherwise the `start` process will fail because there are no files to run just yet.
```

File: packages/hmr/README.md [hunk 21]
```diff
@@ -0,0 +1,179 @@
...
+	// this has to be called after you have logged into your bot
+	if (process.env.NODE_ENV === 'development') start();
+}
+
+void main();
+```
+
+In order for proper HMR, you will need to recompile your code, so you will need to edit your package.json:
+
+- replace the dev and start script
```
Comment (favna) [inline packages/hmr/README.md:101]:
```suggestion
In order for HMR to pick up your compiled JavaScript files, you will need to recompile your code. To that end, we will configure a `dev` script in `package.json` scripts that runs `build` in parallel with `start`:
```

File: packages/hmr/package.json [hunk 22]
```diff
@@ -0,0 +1,68 @@
+{
+	"name": "@sapphire/plugin-hmr",
+	"version": "0.0.1",
+	"description": "Plugin for @sapphire/framework for hot module replacement with commands, listeners and more",
```
Comment (favna) [inline packages/hmr/package.json:4]:
```suggestion
	"description": "Plugin for @sapphire/framework for hot module replacement for pieces",
```

File: packages/hmr/package.json [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+			"import": "./register.mjs",
+			"require": "./register.js"
+		}
+	},
+	"sideEffects": [
+		"./dist/register.js",
+		"./register.js",
+		"./register.mjs"
+	],
+	"homepage": "https://github.com/sapphiredev/plugins/tree/main/packages/hmr",
```
Comment (favna) [inline packages/hmr/package.json:26]:
```suggestion
	"homepage": "https://sapphirejs.dev",
```

File: packages/hmr/src/index.ts [hunk 24]
```diff
@@ -0,0 +1,57 @@
+import { container, Piece, Store } from '@sapphire/framework';
+import { watch } from 'chokidar';
+import { relative } from 'path';
+
+/**
+ * Starts HMR for all registered {@link Store Stores} in {@link container.stores the main container}.
+ */
+export function start() {
+	container.logger.info('HMR is enabled!');
+	for (const store of container.stores.values()) {
```
Comment (favna) [inline packages/hmr/src/index.ts:10]:
```suggestion
	container.logger.info('HMR is enabled!');

	for (const store of container.stores.values()) {
```

File: packages/hmr/src/index.ts [hunk 25]
```diff
@@ -0,0 +1,57 @@
...
+
+	const pieceToDelete = store.find((piece) => piece.location.full === path);
+	if (!pieceToDelete) return;
+
+	try {
+		await pieceToDelete.unload();
+		container.logger.info(`[HMR] Unloaded ${pieceToDelete.name}`);
+	} catch (error) {
+		container.logger.error(`[HMR] Failed to unload ${pieceToDelete.name}`, error);
+	}
```
Comment (favna) [inline packages/hmr/src/index.ts:31]: Seeing as this package depends on framework anyway we might as well use `fromAsync` here

```suggestion
const result = fromAsync(async () => {
	await pieceToDelete.unload();
	container.logger.info(`[HMR] Unloaded ${pieceToDelete.name}`);
});

if (isErr(result)) {
	container.logger.error(`[HMR] Failed to unload ${pieceToDelete.name}`, error);
}
```

File: packages/hmr/src/tsconfig.json [hunk 26]
```diff
@@ -0,0 +1,11 @@
+{
+	"extends": "../../../tsconfig.base.json",
+	"compilerOptions": {
+		"rootDir": "./",
+		"outDir": "../dist",
+		"composite": true,
+		"preserveConstEnums": true,
```
Comment (favna) [inline packages/hmr/src/tsconfig.json:7]:
```suggestion
```

**Review by favna — CHANGES_REQUESTED** [review]

File: packages/hmr/package.json (hunk 22 above)
Comment (vladfrangu) [reply packages/hmr/package.json:4]: I'd argue it should be reload not replacement

File: packages/hmr/package.json [hunk 27]
```diff
@@ -0,0 +1,68 @@
...
+	"scripts": {
+		"lint": "eslint src --ext ts --fix",
+		"build": "tsc -b src && yarn esm:register && yarn esm:default",
+		"esm:register": "gen-esm-wrapper dist/register.js dist/register.mjs",
+		"esm:default": "gen-esm-wrapper dist/index.js dist/index.mjs",
+		"prepublishOnly": "yarn build"
+	},
+	"dependencies": {
+		"chokidar": "^3.5.2",
+		"tslib": "^2.3.1"
```
Comment (vladfrangu) [inline packages/hmr/package.json:36]:
```suggestion
		"tslib": "2.x"
```

File: packages/hmr/src/index.ts [hunk 28]
```diff
@@ -0,0 +1,57 @@
+import { container, Piece, Store } from '@sapphire/framework';
+import { watch } from 'chokidar';
+import { relative } from 'path';
+
+/**
+ * Starts HMR for all registered {@link Store Stores} in {@link container.stores the main container}.
+ */
+export function start() {
+	container.logger.info('HMR is enabled!');
```
Comment (vladfrangu) [inline packages/hmr/src/index.ts:9]: Honestly, for logs from the plugin use a header like `HMR-Plugin: Enabled HMR.` or w/e

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: README.md (hunk 8 above)
Comment (Suyashtnt) [reply README.md:31]: This was done by prettier, not by me

File: README.md (hunk 9 above)
Comment (Suyashtnt) [reply README.md:83]: This was done by prettier, not by me

Comment (favna) [reply README.md:83]: No it wasn't because our prettier config isn't set up like that. Regardless, it doesn't take away that you should revert it.

File: README.md (hunk 8 above)
Comment (favna) [reply README.md:31]: No it wasn't because our prettier config isn't set up like that. Regardless, it doesn't take away that you should revert it.

File: README.md (hunk 9 above)
Comment (Suyashtnt) [reply README.md:83]: I know

**Review by favna — DISMISSED** [review]

File: packages/hmr/README.md [hunk 29]
```diff
@@ -0,0 +1,178 @@
...
+async function main() {
+	await client.login();
+
+	// this has to be called after you have logged into your bot
+	if (process.env.NODE_ENV === 'development') {
+		start();
+	}
+}
+
+main();
```
Comment (kyranet) [inline packages/hmr/README.md:69]:
```suggestion
void main();
```

**Review by kyranet — CHANGES_REQUESTED** [review]

**Review by favna — APPROVED** [review]

**Review by kyranet — DISMISSED** [review]

**Review by vladfrangu — APPROVED** [review]

**Review by favna — APPROVED** [review]

**Review by kyranet — APPROVED** [review]
